src/Payments.sol

```diff
@@ -0,0 +1,612 @@
+     * @param railId The identifier of the rail being settled.
+     * @param rail A storage reference to the rail.
+     * @param epochStart The starting epoch of this settlement segment.
+     * @param epochEnd The ending epoch of this settlement segment.
+     * @param rate The rate for this settlement segment.
+     * @return settledAmount The final amount to settle for this segment.
+     * @return settledUntil The final epoch that effectively got settled.
+     * @return note An optional note returned by the arbitrator.
+     */
+    function _settleSegment(
+        uint256 railId,
+        Rail storage rail,
+        uint256 epochStart,
+        uint256 epochEnd,
+        uint256 rate
+    ) 
+        internal 
+        returns (
+            uint256 settledAmount, 
+            uint256 settledUntil, 
+            string memory note
+        ) 
+    {
+        // Calculate the intended settlement amount for the time span.
+        uint256 duration = epochEnd - epochStart;
+        settledAmount = rate * duration;
+        settledUntil = epochEnd;
+
+        // If the rail has an assigned arbiter, call it for arbitration on this segment.
+        if (rail.arbiter != address(0)) {
+            IArbiter arbiter = IArbiter(rail.arbiter);
+            IArbiter.ArbitrationResult memory arbResult = arbiter.arbitratePayment(
+                railId,
+                settledAmount,
+                epochStart,
+                epochEnd
+            );
+
+            require(arbResult.approved, "failed to settle: arbiter refused payment");
+            require(
+                arbResult.settleUpto <= epochEnd,
+                "failed to settle: arbiter settled beyond segment end"
+            );
+            require(
+                arbResult.modifiedAmount <= rate * (arbResult.settleUpto - epochStart),
+                "failed to settle: arbiter modified amount exceeds maximum for settled duration"
+            );
+
+            // Adjust the amount and the final epoch according to the arbiter's decision.
+            settledUntil = arbResult.settleUpto;
+            settledAmount = arbResult.modifiedAmount;
+            note = arbResult.note;
+        }
+
+        // Revert if no progress was made (i.e., the final epoch did not advance).
+        require(settledUntil > epochStart, "failed to settle: no progress made in settlement");
+
+        return (settledAmount, settledUntil, note);
+    }
+
+    function _settleWithRateChanges(
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 initialRate,
+        uint256 startEpoch,
+        uint256 targetEpoch
+    ) internal returns (uint256 totalSettled, uint256 finalEpoch, string memory note) {
+        totalSettled = 0;
+        uint256 currentEpoch = startEpoch;
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
```

Comment:
currentEpoch < targetEpoch is what we need here. if it is equal, we are done and we should exit the loop.